src/Model/ModelInterface.php

```diff
@@ -10,37 +10,31 @@ interface ModelInterface
     /**
      * Marks the entity for adding.
      *
-     * @param object $entity
-     *
      * @return $this
      */
-    public function add (object $entity);
+    public function add (object $entity) : self;
 
 
     /**
      * Updates the given entity.
      *
-     * @param object $entity
-     *
      * @return $this
      */
-    public function update (object $entity);
+    public function update (object $entity) : self;
 
 
     /**
      * Marks the entity for removal.
      *
-     * @param object $entity
-     *
      * @return $this
      */
-    public function remove (object $entity);
+    public function remove (object $entity) : self;
 
 
     /**
      * Flushes the entity changes to the database.
      *
      * @return $this
      */
-    public function flush ();
+    public function flush () : self;
```

Comment:
This change is wrong, unfortunately.

![2019-11-25 at 12 57](https://user-images.githubusercontent.com/1032411/69538691-2be85400-0f83-11ea-9d55-8d14e75ec14d.png)

We are not returning the static `self`, but rather the dynamic `static` (look up "late static binding" in PHP as reference / explanation). So in my `TestModel` your code says that we are returning a (static determined) `ModelInterface`, although the code actually returns a `TestModel`. You can see that the `hello` method is therefore missing.

The 100% correct code would be sth like `public function flush () : static;` or even `public function flush () : this;`, however both of these are not (yet?) supported in PHP.

Although this change is technically valid, it is semantically invalid, as we lose a lot of information. And having proper auto completion here is more important than some type checks.

tl;dr please remove.